drivers/sensor/scd4x/scd4x.c

```diff
@@ -0,0 +1,586 @@
+		LOG_ERR("Failed to read data from device. (%d)", rc);
+		return rc;
+	}
+
+	k_sleep(K_MSEC(1));
+
+	uint16_t serial_number0 = sys_get_be16(&rx_buf[0]);
+	if (scd4x_compute_crc(serial_number0) != rx_buf[2]) {
+		LOG_ERR("Invalid CRC0 for serial number.");
+		return -EIO;
+	}
+
+	uint16_t serial_number1 = sys_get_be16(&rx_buf[3]);
+	if (scd4x_compute_crc(serial_number1) != rx_buf[5]) {
+		LOG_ERR("Invalid CRC1 for serial number.");
+		return -EIO;
+	}
+
+	uint16_t serial_number2 = sys_get_be16(&rx_buf[6]);
+	if (scd4x_compute_crc(serial_number2) != rx_buf[8]) {
+		LOG_ERR("Invalid CRC2 for serial number.");
+		return -EIO;
+	}
+
+	snprintf(data->serial_number, sizeof(data->serial_number), "0x%04x%04x%04x", serial_number0, serial_number1, serial_number2);
+
+	return rc;
+}
+
+
+/*
+ * Process the measurement returned from the sensor. Response is a 9 byte buffer containing 3 sensor values.
+ * 
+ * Sensor measurement values are 2 bytes long, and each one is followed by a 1 byte CRC calculated by the
+ * sensor.
+ * 
+ * On SCD41 in single shot measurement mode, if only the temperature and humidity channels have been requested
+ * by the user, the sensor will still produce a CO2 value but it will always be 0 ppm.
+ */ 
+static int scd4x_read_sample(const struct device *dev,
+		uint16_t *t_sample,
+		uint16_t *rh_sample,
+		uint16_t *co2_sample)
+{
+
+	const struct scd4x_config *cfg = dev->config;
+	uint8_t rx_buf[9];
+	int rc;
+
+	rc = i2c_read_dt(&cfg->bus, rx_buf, sizeof(rx_buf));
+	if (rc < 0) {
+		LOG_ERR("Failed to read data from device.");
+		return rc;
+	}
+
+	*co2_sample = sys_get_be16(rx_buf);
+	if (scd4x_compute_crc(*co2_sample) != rx_buf[2]) {
+		LOG_ERR("Invalid CRC for CO2.");
+		return -EIO;
+	}
+
+	*t_sample = sys_get_be16(&rx_buf[3]);
+	if (scd4x_compute_crc(*t_sample) != rx_buf[5]) {
+		LOG_ERR("Invalid CRC for T.");
+		return -EIO;
+	}
+
+	*rh_sample = sys_get_be16(&rx_buf[6]);
+	if (scd4x_compute_crc(*rh_sample) != rx_buf[8]) {
+		LOG_ERR("Invalid CRC for RH.");
+		return -EIO;
+	}
+
+	return 0;
+}
+
+
+static int scd4x_sample_fetch(const struct device *dev,
+							  enum sensor_channel chan)
+{
+	const struct scd4x_config *cfg = dev->config;
+	struct scd4x_data *data = dev->data;
+	int rc;
+
+	if (chan != SENSOR_CHAN_ALL &&
+		chan != SENSOR_CHAN_AMBIENT_TEMP &&
+		chan != SENSOR_CHAN_HUMIDITY &&
+		chan != SENSOR_CHAN_CO2) {
+		return -ENOTSUP;
+	}
+	/*
+	 * SCD41 in single shot measure mode. The requested sensor channels determine which command is sent
+	 * because the wait time is different by a factor of 100. The full measurement takes 5000ms while
+	 * the temperature/humidity only command takes 50ms.
+	 */
+	if (cfg->model == SCD41 && cfg->measure_mode == MEASURE_MODE_SINGLE_SHOT) {
+		#if defined(SCD4X_POWER_DOWN_SINGLE_SHOT_MEASUREMENT)
+		/*
+		 * Wake up the sensor if necessary before issuing a single shot command, will be powered
+		 * down again after reading the measurement.
+		 */
+		scd4x_wake_up(dev);
```

Comment:
note: you should be using the PM runtime API, see https://docs.zephyrproject.org/latest/services/pm/device_runtime.html